hazelcast/src/main/java/com/hazelcast/map/impl/MapQueryResultSizeLimitHelper.java

```diff
@@ -0,0 +1,130 @@
+package com.hazelcast.map.impl;
+
+import com.hazelcast.instance.GroupProperties;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.map.QueryResultSizeExceededException;
+import com.hazelcast.spi.NodeEngine;
+
+import java.util.Collection;
+
+/**
+ * Helper class to limit query result size.
+ */
+public class MapQueryResultSizeLimitHelper {
+
+    // due to the unequal data distribution we set a minimum value for the trigger to ensure a better distribution
+    public static final int MINIMUM_MAX_RESULT_LIMIT = 100000;
+
+    // due to the unequal data distribution we add a security margin to prevent false positives
+    public static final float MAX_RESULT_LIMIT_FACTOR = 1.15f;
+
+    // special value to mark the disabled state
+    static final int DISABLED = -1;
+
+    private final MapServiceContext mapServiceContext;
+    private final ILogger log;
+
+    private final int maxResultLimit;
+    private final int maxLocalPartitionsLimitForPreCheck;
+    private final float resultLimitPerPartition;
+
+    private final boolean isQueryResultLimitEnabled;
+    private final boolean isPreCheckEnabled;
+
+    public MapQueryResultSizeLimitHelper(MapServiceContext mapServiceContext, ILogger log) {
+        NodeEngine nodeEngine = mapServiceContext.getNodeEngine();
+
+        this.mapServiceContext = mapServiceContext;
+        this.log = log;
+
+        GroupProperties groupProperties = nodeEngine.getGroupProperties();
+        this.maxResultLimit = getMaxResultLimit(groupProperties);
+        this.maxLocalPartitionsLimitForPreCheck = getMaxLocalPartitionsLimitForPreCheck(groupProperties);
+        this.resultLimitPerPartition = maxResultLimit * MAX_RESULT_LIMIT_FACTOR / getPartitionCount(groupProperties);
+
+        this.isQueryResultLimitEnabled = (maxResultLimit != DISABLED);
+        this.isPreCheckEnabled = (maxLocalPartitionsLimitForPreCheck != DISABLED);
+    }
+
+    public boolean isQueryResultLimitEnabled() {
+        return isQueryResultLimitEnabled;
+    }
+
+    public long getNodeResultLimit(int ownedPartitions) {
+        return (long) Math.ceil(resultLimitPerPartition * ownedPartitions);
+    }
+
+    boolean isPreCheckEnabled() {
+        return isPreCheckEnabled;
+    }
+
+    void checkMaxResultLimitOnLocalPartitions(String mapName) {
```

Comment:
I don't know a good design pattern to split up those "if (check) return; if (check) return;" methods. I extracted one method, but I don't see any obvious possibilities to refactor something else.